File: 347-top-k-frequent-elements/top-k-frequent-elements.py
class Solution(object):
    def topKFrequent(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: List[int]
        """

        counts = Counter(nums)

        max_freq = max([counts[x] for x in counts])
        count_freq = defaultdict(list)

        for num in counts:
            count_freq[counts[num]].append(num)
        
        res = []

        for i in range(max_freq, 0, -1):
            for num in count_freq[i]:
                res.append(num)
                if len(res) == k:
                    return res
        

File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> freq = new HashMap<>();

        for (int num: nums){
            freq.put(num, freq.getOrDefault(num, 0) + 1);
        }

        List<List<Integer>> counts = new ArrayList<>();

        for (int i=0; i<= nums.length; i++){
            counts.add(new ArrayList());
        }

        for (Integer num : freq.keySet()){
            int frequency = freq.get(num);
            counts.get(frequency).add(num);
        }

        int[] res = new int[k];
        int curr = 0;

        for (int i=counts.size()-1; i >= 0 && curr < k; i--){
            for (int num: counts.get(i)){
                res[curr] = num;
                curr += 1;
                if (curr == k){
                    return res;
                }
            }
        }

        return res;
    }
}